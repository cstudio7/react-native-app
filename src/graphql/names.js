import gql from 'graphql-tag';

export const femaleNamesQuery = gql`
  {
    femaleNames {
      id
      name
    }
  }
`;

export const maleNamesQuery = gql`
  {
    maleNames {
      id
      name
    }
  }
`;
