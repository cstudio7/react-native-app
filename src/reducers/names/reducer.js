const R = require('ramda');

import {
  LISTSCREEN_NAME_FAVORITE,
  FAVSCREEN_NAME_FAVORITE,
  LISTSCREEN_NAME_UNFAVORITE,
  FAVSCREEN_NAME_UNFAVORITE,
  FETCH_NAMES_SUCCESS
} from '../../constants/constants';
import createReducer from '../../modules/createReducer/createReducer';

const initialState = {
  female: [],
  male: []
};

const updateItemInArray = (state, name, gender) => {
  const genderData = state[gender];
  const index = R.findIndex(R.propEq('id', name.id))(genderData);

  return {
    ...state,
    [gender]: [
      ...genderData.slice(0, index),
      name,
      ...genderData.slice(index + 1)
    ]
  };
};

const mergeGenderNamesWithState = (state, response) => {
  const gender = response.femaleNames ? 'female' : 'male';
  if (!R.path([gender, 'length'], state)) {
    return {
      ...state,
      [gender]: response.femaleNames || response.maleNames
    };
  }

  return {
    ...state,
    [gender]: (response.femaleNames || response.maleNames).map(name =>
      R.mergeAll([
        name,
        state[gender].find(stateName => stateName.name === name.name)
      ])
    )
  };
};

const saveNames = (state, action) =>
  mergeGenderNamesWithState(state, action.response);

const favoriteName = (state, action) => {
  const { payload, type } = action;
  const newName = {
    ...payload.name,
    isFavorite:
      type === FAVSCREEN_NAME_FAVORITE || type === LISTSCREEN_NAME_FAVORITE
  };
  return updateItemInArray(state, newName, payload.gender);
};

const unfavoriteName = favoriteName;

export default createReducer(initialState, {
  [FETCH_NAMES_SUCCESS]: saveNames,
  [LISTSCREEN_NAME_FAVORITE]: favoriteName,
  [FAVSCREEN_NAME_FAVORITE]: favoriteName,
  [LISTSCREEN_NAME_UNFAVORITE]: unfavoriteName,
  [FAVSCREEN_NAME_UNFAVORITE]: unfavoriteName
});
