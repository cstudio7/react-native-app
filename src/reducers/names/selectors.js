export const getNames = (state, { route, screen }) => {
  const names = state.names[route.key];
  if (screen === 'Favorites') {
    return names.filter(name => name.isFavorite);
  }
  return names;
};