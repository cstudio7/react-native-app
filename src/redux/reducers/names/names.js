const R = require('ramda');

import Actions from '../../../constants/Actions';

export const initialState = {
  female: [],
  male: []
};

const updateName = (name, gender, state) => {
  const genderData = state[gender];
  const index = R.findIndex(R.propEq('id', name.id))(genderData);

  return {
    ...state,
    [gender]: [...genderData.slice(0, index), name, ...genderData.slice(index + 1)]
  };
};

export default (state = initialState, action) => {
  const { type, gender } = action;

  if (type === Actions.saveNames) {
    return {
      ...state,
      [gender]: R.uniq([...state[gender], ...action.names])
    };
  }

  if (type === Actions.favoriteName) {
    return updateName({ ...action.name, isFavorite: true }, gender, state);
  }

  if (type === Actions.unfavoriteName) {
    return updateName({ ...action.name, isFavorite: false }, gender, state);
  }

  return state;
};
