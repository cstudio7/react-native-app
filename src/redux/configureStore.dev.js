import { createStore, applyMiddleware, compose } from 'redux';
import { createLogger } from 'redux-logger';
import rootReducer from './rootReducer/rootReducer';

const configureStore = preloadedState => {
  const store = createStore(
    rootReducer,
    preloadedState,
    compose(applyMiddleware(createLogger()))
  );

  if (module.hot) {
    // Enable Webpack hot module replacement for reducers
    module.hot.accept('./rootReducer/rootReducer', () => {
      const nextRootReducer = require('./rootReducer/rootReducer')
        .default;
      store.replaceReducer(nextRootReducer);
    });
  }

  return store;
};

export default configureStore;
