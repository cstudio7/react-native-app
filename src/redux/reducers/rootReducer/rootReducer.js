import { combineReducers } from 'redux';
import names from '../names/names';

const rootReducer = combineReducers({
  names
});

export default rootReducer;
