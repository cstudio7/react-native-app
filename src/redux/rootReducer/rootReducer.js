import { combineReducers } from 'redux';
import names from '../namesReducer/reducer';

const rootReducer = combineReducers({
  names
});

export default rootReducer;
