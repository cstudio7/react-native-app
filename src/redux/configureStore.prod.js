import { createStore, applyMiddleware, compose } from 'redux';
import { persistStore, persistReducer } from 'redux-persist';
import storage from 'redux-persist/lib/storage'; // defaults to localStorage for web and AsyncStorage for react-native
import amplitude from '../middleware/amplitude/amplitude';
import rootReducer from './rootReducer/rootReducer';

const configureStore = () => {
  const persistConfig = {
    key: 'root',
    storage
  };

  const persistedReducer = persistReducer(persistConfig, rootReducer);
  const store = createStore(
    persistedReducer,
    compose(applyMiddleware(amplitude))
  );
  const persistor = persistStore(store);

  return { store, persistor };
};

export default configureStore;
