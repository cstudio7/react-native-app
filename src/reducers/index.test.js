import { createStore } from 'redux';

import names from './names/names';
import rootReducer from './';

const store = createStore(rootReducer);

it('checks that initial state of the root reducer matches what child reducers return given an empty action', () => {
  expect(store.getState()).toMatchSnapshot();
});