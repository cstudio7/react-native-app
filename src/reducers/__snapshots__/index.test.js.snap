// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`check that initial state of the root reducer matches what child reducers return given an empty action 1`] = `
Object {
  "names": Object {
    "female": Array [],
    "male": Array [],
  },
}
`;

exports[`checks that initial state of the root reducer matches what child reducers return given an empty action 1`] = `
Object {
  "names": Object {
    "female": Array [],
    "male": Array [],
  },
}
`;
